python/lsst/meas/base/pluginsBase.py

```diff
@@ -0,0 +1,100 @@
+    # named class constants for execution order
+    CENTROID_ORDER = 0.0
+    SHAPE_ORDER = 1.0
+    FLUX_ORDER = 2.0
+    APCORR_ORDER = 3.0
+    CLASSIFY_ORDER = 4.0
+    DEFAULT_AFTERBURNER = 5.0
+
+    @classmethod
+    def getExecutionOrder(cls):
+        """Sets the relative order of plugins (smaller numbers run first).
+
+        In general, the following class constants should be used (other values
+        are also allowed, but should be avoided unless they are needed):
+        CENTROID_ORDER      centroids and other algorithms that require only a Footprint and its Peaks as input
+        SHAPE_ORDER         shape measurements and other algorithms that require getCentroid() to return
+                            a good centroid (in addition to a Footprint and its Peaks).
+        FLUX_ORDER          flux algorithms that require both getShape() and getCentroid(),
+                            in addition to a Footprint and its Peaks
+        CLASSIFY_ORDER      algorithms that operate on aperture-corrected fluxes
+        DEFAULT_AFTERBURNER plugins that only operate on the catalog
+
+        Must be reimplemented as a class method by concrete derived classes.
+
+        This approach was chosen instead of a full graph-based analysis of dependencies
+        because algorithm dependencies are usually both quite simple and entirely substitutable:
+        an algorithm that requires a centroid can typically make use of any centroid algorithms
+        outputs.  That makes it relatively easy to figure out the correct value to use for any
+        particular algorithm.
+        """
+        raise NotImplementedError("All plugins must implement getExecutionOrder()")
+
+    def __init__(self, config, name):
+        """!
+        Initialize the plugin object.
+
+        @param[in]  config       An instance of this class's ConfigClass.
+        @param[in]  name         The string the plugin was registered with.
+        """
+        object.__init__(self)
+        self.config = config
+        self.name = name
+
+    def fail(self, measRecord, error=None):
+        """!
+        Record a failure of the measure or measureN() method.
+
+        When the plugin raises an exception, framework will call
+        fail() to allow the plugin to set its failure flag
+        field(s).  When measureN() raises an exception, fail() will be
+        called repeatedly with all the records that were being
+        measured.
+
+        If the exception is a MeasurementError, it will be passed as
+        the error argument; in all other cases the error argument will
+        be None, and the failure will be logged by the measurement
+        framework as a warning.
+        """
+        traceback.print_exc()
+        message = ("The algorithm '%s' thinks it cannot fail, but it did; "
+                   "please report this as a bug (the full traceback is above)."
+                   % self.__class__.__name__)
+        raise NotImplementedError(message)
```

Comment:
This exception is not what is being raised if the plugin fails, each plugin gets to decide what to do in that case. This exception is just to inform that the base class method fail was not overloaded in the child plugin class.